README.md
# Wireshell 0.3.4
**An extendable ProcessWire command line companion**

Aiming for: a command line interface for ProcessWire (like Drush is for Drupal), for running certain (e.g. maintenance, installation) tasks quickly - without having to use the ProcessWire admin Interface.
Since ProcessWire has a powerful API and an easy way of being bootstrapped into CLIs like this, I think such a tool has a certain potential in the PW universe.

It's totally not the first approach of this kind. But: this one is easily extendable, and based on PHP (specifically: the Console component of the Symfony Framework). Every command is tidily wrapped in its own class, dependencies are clearly visible.

## Install Wireshell on your system
Wireshell requires Composer and a local PHP installation >= 5.4.0.

1. Download and install Composer (if it isn't on your system already), globally: https://getcomposer.org/doc/00-intro.md#globally
2. Run `$ composer global require wireshell/wireshell`
3. Add Wireshell to your system path: `export PATH="$HOME/.composer/vendor/bin:$PATH"` in `.bashrc` (or similar) on unix-based systems, and `%appdata%\Composer\vendor\bin` on Windows.
4. You should be able to run `$ wireshell` or `$ php wireshell` now.

## Commands

Available commands in Wireshell:

### New

#### General ProcessWire installation

```
$ wireshell new /path/where/to/install
```

Use `--dev` option for dev branch instead. Further available options (with example values):

```
--dbHost=127.0.0.1 --dbUser=db-user --dbName=db-name --dbPass=db-password --dbEngine=MyISAM --dbPort=3306
--dbCharset=utf8 --httpHosts=pw.dev --adminUrl=processwire --username=admin --userpass=abcd1234
--useremail=someone@example.com --profile=/path/to/myprofile.zip --dev --no-install --chmodDir=775 --chmodFile=644
```

If you don't pass the options, it will ask interactively or use default values.
For MAMP users, provide at least `--dbHost=127.0.0.1 --chmodDir=777`.

#### Custom profile ProcessWire installation π

```
$ wireshell new /path/where/to/install --profile=/path/to/myprofile.zip
```

You can also install profiles. Current structure of zip is as:

```
myprofile/
    site-default/
        modules
        templates
    composer.json
```

#### Download ProcessWire only

```
$ wireshell new /path/where/to/install --no-install
```

Downloads and unzips ProcessWires master branch into current directory.


### User Create

```
$ wireshell user:create {user-name}
```

Creating a user. Available options (with example values):

```
--email=otto@example.org --roles=superuser,editor
```

Use `--roles` for setting one or more user roles (given the supplied role(s) exist). Role `guest` is attached by default.

**Alias:** `$ wireshell u:c`


### Role Create

```
$ wireshell role:create {role-name}
```

Creating a role named editor.

**Alias:** `$ wireshell r:c`


#### Template Create

```
$ wireshell template:create {template-name}
```

Creating a template, and corresponding empty php file in `sites/templates`. Use `--nofile` to prevent file creation. Further available options (with example values):

```
--fields=body,website
```

Field `title` is attached by default.

**Alias:** `$ wireshell c:t`

#### Template Fields

```
$ wireshell template:fields {template-name} --fields={field-name},{field-name}
```

Assign existing field(s) to existing templates, so `--field` option is mandatory here.

**Alias:** `$ wireshell t:f`


#### Field Create

```
$ wireshell field:create {field-name}
```

Creates a text field {field-name}. Available options:

```
--label=Field --desc="Fancy description for field" --type={text|textarea|email|datetime|checkbox|file|float|image|integer|page|url}
```

**Alias:** `$ wireshell f:c`

#### Module Download

```
$ wireshell mod:download {module-name},{module-name}
```

Downloads a module.

**Alias:** `$ wireshell m:dl`


#### Module Enable

```
$ wireshell mod:enable {module-name}
```

Enables and installs a (present!) module.

**Alias:** `$ wireshell m:e`

#### Module Disable

```
 $ wireshell mod:disable {module-name}
```

Disables and uninstalls a module.

Available option:

```
$ wireshell mod:disable --rm {module-name}
```

Disables, uninstalls and deletes a module. 

**Alias:** `$ wireshell m:d`

#### Show Version

```
$ wireshell show:version
```

Outputs the version number of the current ProcessWire installation

**Alias:** `$ wireshell s:v`

#### Show Admin URL

```
$ wireshell show:admin
```

Outputs the admin page url of the current ProcessWire installation

**Alias:** `$ wireshell s:a`

#### Serve

```
$ wireshell serve
```

A wrapper for 'php -S localhost:8000, fires the small PHP web server and lets you bypass the configuration of a virtual host (the database environment must be present, though). Mainly an example for passing through console commands.

### Backup DB

```
$ wireshell backup:db
```

Connects to MySQL database and dumps its complete content into an sql file in the PW installation's root folder. When no file name provided defaulting to a date-and-time based filename.

```
--filename=some_filename
```


## Target group
Comparable to the usage context of (Laravel) Artisan and Drush, I see Wireshell as a tool aiming to help at local development, and aimed at developers who use the console anyway. So if you're developing locally and dealing with many local installations, the tool could help you in speeding things up.

## Current state
Wireshell is still in an early phase. If you encounter bugs, please report them in the project's [GitHub Issues](https://github.com/marcus-herrmann/wireshell/issues). Cheers!

## Technical Background
[The Symfony Console component](http://symfony.com/doc/current/components/console/introduction.html). NewCommand mainly consists of Taylor Otwell's [Laravel Installer](https://github.com/laravel/installer), and partly methods from [Somas PW Online installer](https://github.com/somatonic/PWOnlineInstaller) (moving all PW files a "folder up" after de-zipping the received files from GitHub). Also, big thanks to this PHP Screencast series: [https://laracasts.com/series/how-to-build-command-line-apps-in-php/]

## Potential
See the code, other commands can be added in easily since they are only classes. Symfony and the Console component are written in modern, comprehensible, maintainable PHP. Also, as far as the road-map for ProcessWire 3 goes it will be support Composer. So maybe there could be even more possibilities for Wireshell in the future.

And what made me love Drush in the first place were commands like `drush dl modulename && drush en modulname` (Downloads and installs modules without touching the GUI). I want that for PW too! :)

## Version History

* 0.3.3 Hotfix autoload path
* 0.3.2 Added Show Admin Url Command, added `bin` to composer.json
* 0.3.1 Change readme: Installation via Packagist
* 0.3.0 `NewCommand` now installs PW instead of just downloading it (thanks to @HariKT), added Commands for Fields, Modules, Backup
* 0.2.0 Added Create Template Command, extended Create User Command
* 0.1.0 Initial

## Acknowledgements
[HariKT](https://github.com/harikt) for his big contribution of a real "NewCommand" and command line based installer within Wireshell!

## Feedback please
If you have the time, maybe the slightest need for a tool like this and like to test things out - please grab a copy and go for a test drive with Wireshell and leave feedback in the ProcessWire forum and bugs as GitHub Issues. Thanks!






CHANGELOG.md
# Version 0.3.4

* Updated README
* Added CHANGELOG
* Added Contributor Code of Conduct
* Organized Code

# Version 0.3.3

* Hotfix autoload path

# Version 0.3.2

* Added Show Admin Url Command, added `bin` to composer.json

# Version 0.3.1

* Change readme: Installation via Packagist

# Version 0.3.0

* `NewCommand` now installs PW instead of just downloading it (thanks to @HariKT), added Commands for Fields, Modules, Backup

# Version 0.2.0

* Added Create Template Command, extended Create User Command

# Version 0.1.0

* Initial


src/OneclickCommand.php
<?php
/*
 * This file is part of the Symfony Installer package.
 *
 * https://github.com/symfony/symfony-installer
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Wireshell;

use Distill\Distill;
use Distill\Exception\IO\Input\FileCorruptedException;
use Distill\Exception\IO\Input\FileEmptyException;
use Distill\Exception\IO\Output\TargetDirectoryNotWritableException;
use Distill\Strategy\MinimumSize;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Message\Response;
use GuzzleHttp\Subscriber\Progress\Progress;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Helper\ProgressBar;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Question\Question;
use Symfony\Component\Filesystem\Filesystem;
use Symfony\Component\Process\Process;
use Monolog\Logger;
use Monolog\Handler\StreamHandler;
use Wireshell\Installer;

class OneclickCommand extends Command
{
    /**
     * @var Filesystem
     */
    private $fs;

    private $projectName;

    private $projectDir;

    private $version;

    private $compressedFilePath;

    private $requirementsErrors = array();

    private $installer;

    private $master = 'https://github.com/ryancramerdesign/ProcessWire/archive/master.zip';

    private $dev = 'https://github.com/ryancramerdesign/ProcessWire/archive/dev.zip';

    /**
     * @var OutputInterface
     */
    private $output;

    protected function configure()
    {
        $this
            ->setName('one-click-install')
            ->setDescription('Creates a new processwire project.')
            ->addArgument('directory', InputArgument::REQUIRED, 'Directory where the new project will be created.')
            ->addOption('dbUser', null, InputOption::VALUE_REQUIRED, 'Database user.')
            ->addOption('dbName', null, InputOption::VALUE_REQUIRED, 'Database name.')
            ->addOption('dbPass', null, InputOption::VALUE_REQUIRED, 'Database password.')
            ->addOption('dbHost', null, InputOption::VALUE_REQUIRED, 'Database host.')
            ->addOption('dbPort', null, InputOption::VALUE_REQUIRED, 'Database port.')
            ->addOption('dbEngine', null, InputOption::VALUE_REQUIRED, 'Database engine.')
            ->addOption('dbCharset', null, InputOption::VALUE_REQUIRED, 'Database characterset.')
            ->addOption('timezone', null, InputOption::VALUE_REQUIRED, 'Timezone.')
            ->addOption('chmodDir', null, InputOption::VALUE_REQUIRED, 'Directory mode. Defaults 755.')
            ->addOption('chmodFile', null, InputOption::VALUE_REQUIRED, 'File mode. Defaults 644')
            ->addOption('httpHosts', null, InputOption::VALUE_REQUIRED, 'Hostname without www part.')
            ->addOption('adminUrl', null, InputOption::VALUE_REQUIRED, 'Admin url.')
            ->addOption('username', null, InputOption::VALUE_REQUIRED, 'Admin username.')
            ->addOption('userpass', null, InputOption::VALUE_REQUIRED, 'Admin password.')
            ->addOption('useremail', null, InputOption::VALUE_REQUIRED, 'Admin email address.')
            ->addOption('profile', null, InputOption::VALUE_REQUIRED, 'Default site profile.')
            // ->addArgument('color', null, InputOption::VALUE_NONE, 'Default processwire color.')
            ->addOption('dev', null, InputOption::VALUE_NONE, 'Download dev branch');
        ;
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $this->fs = new Filesystem();
        $directory = rtrim(trim($input->getArgument('directory')), DIRECTORY_SEPARATOR);
        $this->projectDir = $this->fs->isAbsolutePath($directory) ? $directory : getcwd().DIRECTORY_SEPARATOR.$directory;
        $this->projectName = basename($directory);

        $logger = new Logger('name');
        $logger->pushHandler(new StreamHandler("php://output"));
        $this->installer = new Installer($logger, $this->projectDir);

        // $this->version = trim($input->getArgument('version'));
        $this->version = '2.4.0';
        $this->output = $output;

        $profile = $input->getOption('profile');
        $branch = ($input->getOption('dev')) ? $this->dev : $this->master;

        try {
            $this
                ->checkProjectName()
                ->download($branch)
                ->extract()
                ->cleanUp()
                ->extractProfile($profile)
                ->checkProcesswireRequirements();

            $helper = $this->getHelper('question');

            $post = array(
                'dbUser' => '',
                'dbName' => '',
                'dbPass' => '',
                'dbHost' => 'localhost',
                'dbPort' => '3306',
                'dbEngine' => 'MyISAM',
                'dbCharset' => 'utf8',
                'timezone' => "366",
                'chmodDir' => '755',
                'chmodFile' => '644',
                'httpHosts' => ''
            );
            $dbUser = $input->getOption('dbUser');
            if (! $dbUser) {
                $question = new Question('Please enter the database user name : ', 'dbUser');
                $dbUser = $helper->ask($input, $output, $question);
            }
            $post['dbUser'] = $dbUser;

            $dbName = $input->getOption('dbName');
            if (! $dbName) {
                $question = new Question('Please enter the database name : ', 'dbName');
                $dbName = $helper->ask($input, $output, $question);
            }
            $post['dbName'] = $dbName;

            $dbPass = $input->getOption('dbPass');
            if (! $dbPass) {
                $question = new Question('Please enter the database password : ', 'dbPass');
                $question->setHidden(true);
                $question->setHiddenFallback(false);
                $dbPass = $helper->ask($input, $output, $question);
            }
            $post['dbPass'] = $dbPass;

            $dbHost = $input->getOption('dbHost');
            if ($dbHost) {
                $post['dbHost'] = $dbHost;
            }

            $dbPort = $input->getOption('dbPort');
            if ($dbPort) {
                $post['dbPort'] = $dbPort;
            }

            $dbEngine = $input->getOption('dbEngine');
            if ($dbEngine) {
                $post['dbEngine'] = $dbEngine;
            }

            $dbCharset = $input->getOption('dbCharset');
            if ($dbCharset) {
                $post['dbCharset'] = $dbCharset;
            }

            $timezone = $input->getOption('timezone');
            if ($timezone) {
                $post['timezone'] = $timezone;
            }

            $chmodDir = $input->getOption('chmodDir');
            if ($chmodDir) {
                $post['chmodDir'] = $chmodDir;
            }

            $chmodFile = $input->getOption('chmodFile');
            if ($chmodFile) {
                $post['chmodFile'] = $chmodFile;
            }

            $httpHosts = $input->getOption('httpHosts');
            if (! $httpHosts) {
                $question = new Question('Please enter the hostname without www. Eg: pw.dev : ', 'httpHosts');
                $httpHosts = $helper->ask($input, $output, $question);
            }
            $post['httpHosts'] = $httpHosts . "\n" . "www.". $httpHosts;

            $accountInfo = array(
                'admin_name' => 'processwire',
                'username' => '',
                'userpass' => '',
                'userpass_confirm' => '',
                'useremail' => '',
                'color' => 'classic',
            );

            $adminUrl = $input->getOption('adminUrl');
            if ($adminUrl) {
                $accountInfo['admin_name'] = $adminUrl;
            }

            $username = $input->getOption('username');
            if (! $username) {
                $question = new Question('Please enter admin user name : ', 'username');
                $username = $helper->ask($input, $output, $question);
            }
            $accountInfo['username'] = $username;

            $userpass = $input->getOption('userpass');
            if (! $userpass) {
                $question = new Question('Please enter admin password : ', 'password');
                $question->setHidden(true);
                $question->setHiddenFallback(false);
                $userpass = $helper->ask($input, $output, $question);
            }
            $accountInfo['userpass'] = $userpass;
            $accountInfo['userpass_confirm'] = $userpass;

            $useremail = $input->getOption('useremail');
            if (! $useremail) {
                $question = new Question('Please enter admin email address : ', 'useremail');
                $useremail = $helper->ask($input, $output, $question);
            }
            $accountInfo['useremail'] = $useremail;

/*
        $colors = array(
            'classic',
            'warm',
            'modern',
            'futura'
        );
*/
            $this
                ->installProcessWire($post, $accountInfo)
            ;
        } catch (\Exception $e) {
            $this->cleanUp();
            throw $e;
        }
    }

    /**
     * Checks whether it's safe to create a new project for the given name in the
     * given directory.
     *
     * @return NewCommand
     *
     * @throws \RuntimeException if a project with the same does already exist
     */
    private function checkProjectName()
    {
        if (is_dir($this->projectDir) && !$this->isEmptyDirectory($this->projectDir)) {
            throw new \RuntimeException(sprintf(
                "There is already a '%s' project in this directory (%s).\n".
                "Change your project name or create it in another directory.",
                $this->projectName, $this->projectDir
            ));
        }

        return $this;
    }

    /**
     * Chooses the best compressed file format to download (ZIP or TGZ) depending upon the
     * available operating system uncompressing commands and the enabled PHP extensions
     * and it downloads the file.
     *
     * @return NewCommand
     *
     * @throws \RuntimeException if the Processwire archive could not be downloaded
     */
    private function download($branch)
    {
        $this->output->writeln("\n Downloading processwire...");

        // decide which is the best compressed version to download
        $distill = new Distill();
        // $branch = 'https://github.com/ryancramerdesign/ProcessWire/archive/master.zip';
        $pwArchiveFile = $distill
            ->getChooser()
            ->setStrategy(new MinimumSize())
            ->addFile($branch)
            // ->addFilesWithDifferentExtensions($branch, ['zip', 'tgz'])
            ->getPreferredFile()
        ;

        /** @var ProgressBar|null $progressBar */
        $progressBar = null;
        $downloadCallback = function ($size, $downloaded, $client, $request, Response $response) use (&$progressBar) {
            // Don't initialize the progress bar for redirects as the size is much smaller
            if ($response->getStatusCode() >= 300) {
                return;
            }

            if (null === $progressBar) {
                ProgressBar::setPlaceholderFormatterDefinition('max', function (ProgressBar $bar) {
                    return $this->formatSize($bar->getMaxSteps());
                });
                ProgressBar::setPlaceholderFormatterDefinition('current', function (ProgressBar $bar) {
                    return str_pad($this->formatSize($bar->getStep()), 11, ' ', STR_PAD_LEFT);
                });

                $progressBar = new ProgressBar($this->output, $size);
                $progressBar->setFormat('%current%/%max% %bar%  %percent:3s%%');
                $progressBar->setRedrawFrequency(max(1, floor($size / 1000)));
                $progressBar->setBarWidth(60);

                if (!defined('PHP_WINDOWS_VERSION_BUILD')) {
                    $progressBar->setEmptyBarCharacter('░'); // light shade character \u2591
                    $progressBar->setProgressCharacter('');
                    $progressBar->setBarCharacter('▓'); // dark shade character \u2593
                }

                $progressBar->start();
            }

            $progressBar->setProgress($downloaded);
        };

        $client = new Client();
        $client->getEmitter()->attach(new Progress(null, $downloadCallback));

        // store the file in a temporary hidden directory with a random name
        $this->compressedFilePath = getcwd().DIRECTORY_SEPARATOR.'.'.uniqid(time()).DIRECTORY_SEPARATOR.'pw.'.pathinfo($pwArchiveFile, PATHINFO_EXTENSION);

        try {
            $response = $client->get($pwArchiveFile);
        } catch (ClientException $e) {
            if ($e->getCode() === 403 || $e->getCode() === 404) {
                throw new \RuntimeException(sprintf(
                    "The selected version (%s) cannot be installed because it does not exist.\n".
                    "Try the special \"latest\" version to install the latest stable processwire release:\n".
                    '%s %s %s latest',
                    $this->version,
                    $_SERVER['PHP_SELF'],
                    $this->getName(),
                    $this->projectDir
                ));
            } else {
                throw new \RuntimeException(sprintf(
                    "The selected version (%s) couldn't be downloaded because of the following error:\n%s",
                    $this->version,
                    $e->getMessage()
                ));
            }
        }

        $this->fs->dumpFile($this->compressedFilePath, $response->getBody());

        if (null !== $progressBar) {
            $progressBar->finish();
            $this->output->writeln("\n");
        }

        return $this;
    }

    /**
     * Extracts the compressed Symfony file (ZIP or TGZ) using the
     * native operating system commands if available or PHP code otherwise.
     *
     * @return NewCommand
     *
     * @throws \RuntimeException if the downloaded archive could not be extracted
     */
    private function extract()
    {
        $this->output->writeln(" Preparing project...\n");

        try {
            $distill = new Distill();
            $extractionSucceeded = $distill->extractWithoutRootDirectory($this->compressedFilePath, $this->projectDir);
        } catch (FileCorruptedException $e) {
            throw new \RuntimeException(sprintf(
                "Processwire can't be installed because the downloaded package is corrupted.\n".
                "To solve this issue, try installing Processwire again.\n%s",
                $this->getExecutedCommand()
            ));
        } catch (FileEmptyException $e) {
            throw new \RuntimeException(sprintf(
                "Processwire can't be installed because the downloaded package is empty.\n".
                "To solve this issue, try installing Processwire again.\n%s",
                $this->getExecutedCommand()
            ));
        } catch (TargetDirectoryNotWritableException $e) {
            throw new \RuntimeException(sprintf(
                "Processwire can't be installed because the installer doesn't have enough\n".
                "permissions to uncompress and rename the package contents.\n".
                "To solve this issue, check the permissions of the %s directory and\n".
                "try installing Processwire again.\n%s",
                getcwd(), $this->getExecutedCommand()
            ));
        } catch (\Exception $e) {
            throw new \RuntimeException(sprintf(
                "Processwire can't be installed because the downloaded package is corrupted\n".
                "or because the installer doesn't have enough permissions to uncompress and\n".
                "rename the package contents.\n".
                "To solve this issue, check the permissions of the %s directory and\n".
                "try installing Processwire again.\n%s",
                getcwd(), $this->getExecutedCommand()
            ));
        }

        if (!$extractionSucceeded) {
            throw new \RuntimeException(
                "Processwire can't be installed because the downloaded package is corrupted\n".
                "or because the uncompress commands of your operating system didn't work."
            );
        }

        return $this;
    }

    /**
     * Removes all the temporary files and directories created to
     * download the project and removes Processwire-related files that don't make
     * sense in a proprietary project.
     *
     * @return NewCommand
     */
    private function cleanUp()
    {
        $this->fs->remove(dirname($this->compressedFilePath));

        try {
            $licenseFile = array($this->projectDir.'/LICENSE');
            $upgradeFiles = glob($this->projectDir.'/UPGRADE*.md');
            $changelogFiles = glob($this->projectDir.'/CHANGELOG*.md');

            $filesToRemove = array_merge($licenseFile, $upgradeFiles, $changelogFiles);
            $this->fs->remove($filesToRemove);

            $readmeContents = sprintf("%s\n%s\n\nA Processwire project created on %s.\n", $this->projectName, str_repeat('=', strlen($this->projectName)), date('F j, Y, g:i a'));
            $this->fs->dumpFile($this->projectDir.'/README.md', $readmeContents);
        } catch (\Exception $e) {
            // don't throw an exception in case any of the Processwire-related files cannot
            // be removed, because this is just an enhancement, not something mandatory
            // for the project
        }

        return $this;
    }

    /**
     * Checks if environment meets processwire requirements
     *
     * @return OneclickCommand
     */
    private function checkProcesswireRequirements()
    {
        $this->installer->compatibilityCheck();
        return $this;
    }

    private function installProcessWire($post, $accountInfo)
    {
        $this->installer->dbSaveConfig($post, $accountInfo);
        return $this;
    }

    /**
     * Utility method to show the number of bytes in a readable format.
     *
     * @param int     $bytes The number of bytes to format
     *
     * @return string The human readable string of bytes (e.g. 4.32MB)
     */
    private function formatSize($bytes)
    {
        $units = array('B', 'KB', 'MB', 'GB', 'TB');

        $bytes = max($bytes, 0);
        $pow = $bytes ? floor(log($bytes, 1024)) : 0;
        $pow = min($pow, count($units) - 1);

        $bytes /= pow(1024, $pow);

        return number_format($bytes, 2).' '.$units[$pow];
    }

    /**
     * Formats the error message contained in the given Requirement item
     * using the optional line length provided.
     *
     * @param \Requirement $requirement The Processwire requirements
     * @param int          $lineSize    The maximum line length
     *
     * @return string
     */
    private function getErrorMessage(\Requirement $requirement, $lineSize = 70)
    {
        if ($requirement->isFulfilled()) {
            return;
        }

        $errorMessage  = wordwrap($requirement->getTestMessage(), $lineSize - 3, PHP_EOL.'   ').PHP_EOL;
        $errorMessage .= '   > '.wordwrap($requirement->getHelpText(), $lineSize - 5, PHP_EOL.'   > ').PHP_EOL;

        return $errorMessage;
    }

    /**
     * Returns the executed command.
     *
     * @return string
     */
    private function getExecutedCommand()
    {
        $version = '';
        if ('latest' !== $this->version) {
            $version = $this->version;
        }

        $pathDirs = explode(PATH_SEPARATOR, $_SERVER['PATH']);
        $executedCommand = $_SERVER['PHP_SELF'];
        $executedCommandDir = dirname($executedCommand);

        if (in_array($executedCommandDir, $pathDirs)) {
            $executedCommand = basename($executedCommand);
        }

        return sprintf('%s new %s %s', $executedCommand, $this->projectName, $version);
    }

    /**
     * Checks whether the given directory is empty or not.
     *
     * @param  string  $dir the path of the directory to check
     * @return bool
     */
    private function isEmptyDirectory($dir)
    {
        // glob() cannot be used because it doesn't take into account hidden files
        // scandir() returns '.'  and '..'  for an empty dir
        return 2 === count(scandir($dir.'/'));
    }

    private function extractProfile($profile)
    {
        if (! $profile) {
            return $this;
        }
        $this->output->writeln(" Extracting profile...\n");

        try {
            $distill = new Distill();
            $extractPath = getcwd().DIRECTORY_SEPARATOR.'.'.uniqid(time()).DIRECTORY_SEPARATOR.'pwprofile';
            $extractionSucceeded = $distill->extractWithoutRootDirectory($profile, $extractPath);
            if ($extractionSucceeded) {
                try {
                    $this->fs->mirror($extractPath, $this->projectDir . '/');
                } catch (\Exception $e) {
                }
                // cleanup
                $this->fs->remove($extractPath);
                try {
                    $process = new Process("cd $this->projectDir; composer install");
                    $process->run(function ($type, $buffer) {
                        if (Process::ERR === $type) {
                            echo ' '.$buffer;
                        } else {
                            echo ' '.$buffer;
                        }
                    });
                } catch (\Exception $e) {
                }
            }
        } catch (FileCorruptedException $e) {
            throw new \RuntimeException(
                "The profile can't be installed because the downloaded package is corrupted.\n"
            );
        } catch (FileEmptyException $e) {
            throw new \RuntimeException(
                "The profile can't be installed because the downloaded package is empty.\n"
            );
        } catch (TargetDirectoryNotWritableException $e) {
            throw new \RuntimeException(
                "The profile can't be installed because the installer doesn't have enough\n".
                "permissions to uncompress and rename the package contents.\n"
            );
        } catch (\Exception $e) {
            throw new \RuntimeException(
                "The profile can't be installed because the downloaded package is corrupted\n".
                "or because the installer doesn't have enough permissions to uncompress and\n".
                "rename the package contents.\n".
                $e->getMessage()
            );
        }

        if (!$extractionSucceeded) {
            throw new \RuntimeException(
                "The profile can't be installed because the downloaded package is corrupted\n".
                "or because the uncompress commands of your operating system didn't work."
            );
        }

        return $this;
    }
}
